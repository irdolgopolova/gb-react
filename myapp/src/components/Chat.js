import {
    TextField,
    Button,
} from "@mui/material";
import React from 'react';

export function Chat({
    onAddMessage,
    newMessage,
    handleChange,
    messages,
    isEmptyMessagesList,
    profileName
}) {
    const renderMessageList = () => {
        if (isEmptyMessagesList) {
            return [];
        }

        return messages.map(message => (
            <div
                key={message.id}
                className={`App-message-list__row ${message.author === profileName ? 'right' : ''}`}
            >
                <p className={`App-message-list ${message.author === profileName ? 'user' : ''}`}>
                    {message.text}
                </p>
            </div>
        ));
    }

    return (
        <>
        <div className="App-message-list__history">
            {renderMessageList()}
        </div>

        <form className="App-form" name="myForm" onSubmit={onAddMessage}>
            <TextField
                id="standard-basic"
                label="Ваше сообщение"
                variant="standard"
                value={newMessage}
                onChange={handleChange}
            />
            <Button onClick={onAddMessage}>Отправить</Button>
        </form>
        </>
    );
}
