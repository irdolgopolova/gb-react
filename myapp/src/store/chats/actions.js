import { db } from "../../services/firebase";

export const ADD_CHAT = "CHATS::ADD_CHAT";
export const CHANGE_CHAT = "CHATS::CHANGE_CHAT";
export const DELETE_CHAT = "CHATS::DELETE_CHAT";

export const addChat = (name) => ({
    type: ADD_CHAT,
    name,
});

export const deleteChat = (chatId) => ({
    type: DELETE_CHAT,
    chatId,
})

const getPayloadFromSnapshot = (snapshot) => {
    const chats = [snapshot.val()];


    return { chats};
}

export const addChatWithFirebase = (chat) => async() => {
    db.ref("chats")
        .child(chat.id)
        .set(chat);
}

export const initChatTracking = () => (dispatch) => {
    db.ref("chats")
        .on("child_changed", (snapshot) => {
            const payload = snapshot.val();

            dispatch({
                type: CHANGE_CHAT,
                payload,
            });
        });

    db.ref("chats")
        .on("child_added", (snapshot) => {
            const payload = snapshot.val();

            dispatch({
                type: CHANGE_CHAT,
                payload,
            });
        });
};
