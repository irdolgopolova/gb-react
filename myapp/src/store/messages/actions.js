export const ADD_MESSAGE = "MESAGES::ADD_MESSAGE";

export const addMessage = (chatId, author, message) => ({
    type: ADD_MESSAGE,
    chatId,
    author,
    message,
})
