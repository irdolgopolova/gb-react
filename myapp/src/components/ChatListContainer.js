import React, { useState } from 'react';
 import { useDispatch, useSelector } from "react-redux";
 import { Navigate, useParams } from "react-router-dom";
 import { addChat, deleteChat } from "../store/chats/actions";
 import { getChatList } from "../store/chats/selectors";
 import { ChatList } from "./ChatList";

 export default function ChatListContainer({ authed }) {
     const { chatId } = useParams();

     const [visible, setVisible] = useState(false);
     const [newChatName, setNewChatName] = useState("");

     const chatList = useSelector(getChatList);

     const dispatch = useDispatch();

     const handleClose = () => setVisible(false);
     const handleOpen = () => setVisible(true);
     const handleChange = (e) => setNewChatName(e.target.value);

     const onAddChat = () => {
         dispatch(addChat(newChatName));
         setNewChatName("");
         handleClose();
     };

     const onDeleteChat = (e) => {
         dispatch(deleteChat(e.target.dataset.chat_id));
     };

     const existsActiveChat = () => {
         return !!chatId
             && chatList.findIndex(i => i.id === Number(chatId)) === -1;
     }

     if (!authed) {
        return (<Navigate to="/"/>);
    }

     return (
         <ChatList
             chatId={chatId}
             chatList={chatList}
             visible={visible}
             handleOpen={handleOpen}
             handleClose={handleClose}
            handleChange={handleChange}
            newChatName={newChatName}
            onAddChat={onAddChat}
            onDeleteChat={onDeleteChat}
            existsActiveChat={existsActiveChat()}
        />
    );
}