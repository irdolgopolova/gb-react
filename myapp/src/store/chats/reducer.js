import { ADD_CHAT, DELETE_CHAT } from "./actions"

const initialState = {
    chatList: [
        {
            id: 1,
            name: 'Бот',
        },
        {
            id: 2,
            name: 'Иванов Иван',
            avatar: '/static/image/avatar/2.jpg'
        },
        {
            id: 3,
            name: 'Петров Петр',
            avatar: '/static/image/avatar/3.jpg'
        },
        {
            id: 4,
            name: 'Попова Елена',
            avatar: '/static/image/avatar/4.jpg'
        }
    ],
}

export const chatsReducer = (state = initialState, action) => {
    switch (action.type) {
        case ADD_CHAT:
            let lastChat = state.chatList[state.chatList.length - 1];
            let newId = lastChat ? lastChat.id + 1 : 1;

            return {
                ...state,
                chatList: [
                    ...state.chatList,
                    {
                        id: newId,
                        name: action.name,
                    },
                ]
            }
        case DELETE_CHAT:
            let index = state.chatList.findIndex(el =>
                Number(el.id) === Number(action.chatId)
            );

            if (index != -1) {
                state.chatList.splice(index, 1);
            }

            return {
                ...state,
                chatList: [
                    ...state.chatList
                ]
            }
        default: return state;
    }
}
