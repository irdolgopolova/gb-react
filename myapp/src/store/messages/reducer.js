import { ADD_MESSAGE, CHANGE_MESSAGES } from "./actions";

const initialState = {
    messages: {},
};

export const messagesReducer = (state = initialState, action) => {
    switch (action.type) {
        case ADD_MESSAGE: {
            const currentList =  state.messages[action.chatId] === undefined
                ? []
                : state.messages[action.chatId];

            return {
                ...state,
                messages: {
                    ...state.messages,
                    [action.chatId]: [
                        ...currentList,
                        {
                            id: `${action.chatId}${currentList.length}`,
                            author: action.author,
                            text: action.message
                        }
                    ]
                }
            }
        }
        case CHANGE_MESSAGES: {
            return ({
                ...state,
                messages: {
                    ...state.messages,
                    [action.payload.chatId]: action.payload.messages,
                },
            });
        }
        default:
            return state;
    }
}
