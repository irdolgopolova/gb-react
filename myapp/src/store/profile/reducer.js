import {SHOW_NAME} from "./actions"

const initialState = {
    showName: false,
    name: 'Default'
}

export const profileReducer = (state = initialState, action) => {
    switch (action.type) {
        case SHOW_NAME:
            return {
                ...state,
                showName: !state.showName
            }
        default:
            return state
    }
}
