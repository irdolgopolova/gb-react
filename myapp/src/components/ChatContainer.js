import React, { useState, useCallback, useEffect } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux'
import { useParams } from "react-router-dom";
import { addMessageWithFirebase, initMessageTracking } from "../store/messages/actions";
import { getProfileName } from "../store/profile/selectors";
import { Chat } from './Chat';

export function ChatContainer() {
    const { chatId } = useParams();

    const [newMessage, setNewMessage] = useState('');
    const messageList = useSelector((state) => state.messages.messages);
    const messages = messageList[chatId];

    const profileName = useSelector(getProfileName, shallowEqual);

    const dispatch = useDispatch();

    const onAddMessage = useCallback((event) => {
        event.preventDefault();

        let newId = (messages !== undefined)
            ? messages.length + 1
            : 1;

        dispatch(
            addMessageWithFirebase(chatId, {
                id: newId,
                author: profileName,
                message: newMessage
            })
        )

        setNewMessage("");
        document.myForm.reset();
    }, [dispatch, newMessage, chatId, messages, profileName]);

    useEffect(() => {
        dispatch(initMessageTracking());
    }, [dispatch]);

    const handleChange = useCallback((e) => {
        setNewMessage(e.target.value);
    }, []);

    const isEmptyMessagesList = (messages) => {
        return messages === undefined ||
            Object.keys(messages).length === 0;
    }

    return (
        <Chat
            newMessage={newMessage}
            profileName={profileName}
            messages={messages}
            isEmptyMessagesList={isEmptyMessagesList(messages)}
            onAddMessage={onAddMessage}
            handleChange={handleChange}
        />
    );
}