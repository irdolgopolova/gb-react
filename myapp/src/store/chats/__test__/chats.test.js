import createMockStore from "redux-mock-store";
import ChatListContainer from "../../../components/ChatListContainer";
import { ADD_CHAT, DELETE_CHAT } from "../actions";
import { chatsReducer, initialState } from "../reducer";
import { Chat } from "./../../../components/Chat";
const { render } = require("@testing-library/react");
const { Provider } = require("react-redux");

describe('Проверка компонента Chats', () => {
    it("Тестирование компонента-контейнера", () => {
        const mockStore = createMockStore();
        const component = render(
            <Provider store={mockStore({ chats: { chatList: []} })}>
                <ChatListContainer authed={true} />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });

    it("Тестирование презентационного компонента", () => {
        const mockStore = createMockStore();
        const component = render(
            <Provider store={mockStore({ messages: [] })}>
                <Chat
                    newMessage={"new message"}
                    profileName={'Default'}
                    messages={[]}
                    isEmptyMessagesList={true}
                />
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });

    it('Тестирование chatsReducer ADD_CHAT', async() => {
        const action = {
            type: ADD_CHAT,
            name: "Пользователь 5"
        };

        expect(chatsReducer(initialState, action)).toEqual({
            ...initialState,
            chatList: [
                ...initialState.chatList,
                {
                    id: 5,
                    name: 'Пользователь 5'
                }
            ]
        });
    });

    it('Тестирование chatsReducer DELETE_CHAT', async() => {
        const action = {
            type: DELETE_CHAT,
            chatId: 4
        };

        expect(chatsReducer(initialState, action)).toEqual({
            ...initialState,
            chatList: [
                {
                    id: 1,
                    name: 'Бот',
                },
                {
                    id: 2,
                    name: 'Иванов Иван',
                    avatar: '/static/image/avatar/2.jpg'
                },
                {
                    id: 3,
                    name: 'Петров Петр',
                    avatar: '/static/image/avatar/3.jpg'
                },
            ]
        });
    });
});