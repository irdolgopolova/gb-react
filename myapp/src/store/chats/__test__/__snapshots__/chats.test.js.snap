// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Проверка компонента Chats Тестирование компонента-контейнера 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App-main__card"
      >
        <div
          class="App-message-list__chat"
        >
          <div
            class="App-message-list__chat_list"
          >
            <ul
              class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
            />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium add-chat css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Добавить чат
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App-main__card"
    >
      <div
        class="App-message-list__chat"
      >
        <div
          class="App-message-list__chat_list"
        >
          <ul
            class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
          />
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium add-chat css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Добавить чат
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Проверка компонента Chats Тестирование презентационного компонента 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App-message-list__history"
      />
      <form
        class="App-form"
        name="myForm"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="standard-basic"
            id="standard-basic-label"
          >
            Ваше сообщение
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="standard-basic"
              type="text"
              value="new message"
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Отправить
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <div
      class="App-message-list__history"
    />
    <form
      class="App-form"
      name="myForm"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="standard-basic"
          id="standard-basic-label"
        >
          Ваше сообщение
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="standard-basic"
            type="text"
            value="new message"
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Отправить
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
